### DISCUSSION TITLE
Adding pH into the model

### DISCUSSION THREAD

**Discussion by iquasere** [nufeb/NUFEB-2#11] [Q&A]
Hey there!

From all examples of NUFEB, I only found the effects of pH modelled in the [Gogulancea paper](doi.org/10.3389/fmicb.2019.01871). By reading it, I understood that the effects of pH were based on speciation of different metabolites important for metabolism, and that the different concentrations of chemical species are then estimated based on temperature. Was temperature fixed at a specific value?

I am trying to simulate the effects of a graphene sheet on pH, and wanted to experiment using localized pH (a lower pH closer to the sheet). Would I be able to simulate the effects on pH by providing different concentrations of chemical species closer to the sheet region? Or is there a more NUFEB way of doing it?

Looking at the [input script](https://github.com/nufeb/NUFEB/blob/master/examples/PAPER-CHEMICAL-IBM/anaerobic-digestion/base/Inputscript.lammps) of Gogulancea, it includes the ```fix kinetics/ph``` command, but I don't see this file on the src of NUFEB, so I assume it's not yet included in NUFEB-dev. Please correct me if I am wrong, @shelllbw.

**Comment by iquasere** [comment]
I'm going to try converting the [```fix_bio_kinetics_ph```](https://github.com/nufeb/NUFEB/blob/master/src/USER-NUFEB/fix_bio_kinetics_ph.cpp) class to NUFEB-dev nomenclature. I see you did that with the [```fix_bio_kinetics_monod```](https://github.com/nufeb/NUFEB/blob/master/src/USER-NUFEB/fix_bio_kinetics_monod.cpp), and both relied on the ```bio``` grid before, so it might be the best to use as example when converting from that to the LAMMPS ```grid``` space.

**Comment by shelllbw** [comment]
The energy model described in Gogulancea, et al (fix_bio_kinetics_ph, fix_bio_kinetics_thermo, and fix_bio_kinetics_energy) is not implemented in NUFEB-dev yet, but I am working on it right now. The migration is not that straightforward, will try to get it done by the end of this month.

Bowen

**Comment by iquasere** [comment]
@shelllbw , is [activity](https://github.com/nufeb/NUFEB/blob/master/src/USER-NUFEB/fix_bio_kinetics.h#L59) of NUFEB's kinetics reimplemented in NUFEB-dev? I see pH was [based on it](https://github.com/nufeb/NUFEB/blob/master/src/USER-NUFEB/fix_bio_kinetics_ph.cpp#L273). I might try to hack it back into grid if it's less complicated than the energy model, likely just recreating the ```activity``` vector like all other [grid components](https://github.com/nufeb/NUFEB-dev/blob/master/src/grid.cpp#L56-L64)?

**Reply by shelllbw** [reply]
Not yet. The current version of energy model in -dev is based on the normal substrate concentration without considering their activities. For the implementation, my idea is instead of introducing `activity` vector in grid.h, we assign the activity value (calculated in fix_ph) to grid->conc directly, so that it can be used in other fixes without altering their codes.

**Reply by iquasere** [reply]
That will mean converting all uses of ```grid->conc```, right? Makes sense to use it instead of polluting the ```grid``` namespace. However, it will mean rewriting all uses of the ```conc``` object... I don't have the capacity for that, will have to wait for its implementation by the main dev x)
